handler/oauth2/flow_generic_code_token.go

```diff
@@ -0,0 +1,244 @@
+	responder.SetAccessToken(accessToken)
+	responder.SetTokenType("bearer")
+	responder.SetExpiresIn(getExpiresIn(requester, fosite.AccessToken, lifeSpan, time.Now().UTC()))
+	responder.SetScopes(requester.GetGrantedScopes())
+	if refreshToken != "" {
+		responder.SetExtra("refresh_token", refreshToken)
+	}
+
+	if err = storage.MaybeCommitTx(ctx, c.CoreStorage); err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return nil
+}
+
+func (c *GenericCodeTokenEndpointHandler) HandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(ctx, requester) {
+		return errorsx.WithStack(errorsx.WithStack(fosite.ErrUnknownRequest))
+	}
+
+	if err := c.ValidateGrantTypes(ctx, requester); err != nil {
+		return err
+	}
+
+	code, _, ar, err := c.GetCodeAndSession(ctx, requester)
+	if err != nil {
+		switch {
+		case errors.Is(err, fosite.ErrInvalidatedAuthorizeCode), errors.Is(err, fosite.ErrInvalidatedDeviceCode):
+			if ar == nil {
+				return fosite.ErrServerError.
+					WithHint("Misconfigured code lead to an error that prohibited the OAuth 2.0 Framework from processing this request.").
+					WithDebug("getCodeSession must return a value for \"fosite.Requester\" when returning \"ErrInvalidatedAuthorizeCode\" or \"ErrInvalidatedDeviceCode\".")
+			}
+
+			return c.revokeTokens(ctx, requester.GetID())
+		case errors.Is(err, fosite.ErrAuthorizationPending):
+			return err
+		case errors.Is(err, fosite.ErrPollingRateLimited):
+			return errorsx.WithStack(err)
+		case errors.Is(err, fosite.ErrNotFound):
+			return errorsx.WithStack(fosite.ErrInvalidGrant.WithWrap(err).WithDebug(err.Error()))
+		default:
+			return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+		}
+	}
+
+	if err = c.ValidateCode(ctx, requester, code); err != nil {
+		return errorsx.WithStack(err)
+	}
+
+	// Override scopes
+	requester.SetRequestedScopes(ar.GetRequestedScopes())
+
+	// Override audiences
+	requester.SetRequestedAudience(ar.GetRequestedAudience())
+
+	// The authorization server MUST ensure that
+	// the authorization code was issued to the authenticated confidential client,
+	// or if the client is public, ensure that the code was issued to "client_id" in the request
+	if ar.GetClient().GetID() != requester.GetClient().GetID() {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.WithHint("The OAuth 2.0 Client ID from this request does not match the one from the authorize request."))
+	}
+
+	forcedRedirectURI := ar.GetRequestForm().Get("redirect_uri")
+	requestedRedirectURI := requester.GetRequestForm().Get("redirect_uri")
+	if forcedRedirectURI != "" && forcedRedirectURI != requestedRedirectURI {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.WithHint("The \"redirect_uri\" from this request does not match the one from the authorize request."))
+	}
```

Comment:
This doesn't really make sense in the case of the device grant. Should we add a call to a generic method, eg `ValidateRequest(context.Context, fosite.AccessRequester, fosite.Requester) error`, that will handle such checks?